src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioProjectOptionsProcessor.cs

```diff
@@ -16,51 +16,77 @@ internal class VisualStudioProjectOptionsProcessor : IDisposable
+        /// as a low cost (in terms of memory) way to determine if the command line
+        /// actually changes and we need to make any downstream updates.
+        /// </summary>
+        private Checksum _commandLineChecksum;
+
+        /// <summary>
+        /// To save space in the managed heap, we dump the entire command-line string to our
+        /// temp-storage-service. This is helpful as compiler command-lines can grow extremely large
+        /// (especially in cases with many references).
+        /// </summary>
+        private ITemporaryStreamStorage _commandLineStorage;
+
         private CommandLineArguments _commandLineArgumentsForCommandLine;
         private string _explicitRuleSetFilePath;
         private IReferenceCountedDisposable<ICacheEntry<string, IRuleSetFile>> _ruleSetFile = null;
 
-        public VisualStudioProjectOptionsProcessor(VisualStudioProject project, HostWorkspaceServices workspaceServices)
+        public VisualStudioProjectOptionsProcessor(
+            VisualStudioProject project,
+            HostWorkspaceServices workspaceServices)
         {
             _project = project ?? throw new ArgumentNullException(nameof(project));
             _workspaceServices = workspaceServices;
             _commandLineParserService = workspaceServices.GetLanguageServices(project.Language).GetRequiredService<ICommandLineParserService>();
+            _temporaryStorageService = workspaceServices.GetService<ITemporaryStorageService>();
+            _commandLineStorage = _temporaryStorageService.CreateTemporaryStreamStorage();
 
-            // Set up _commandLineArgumentsForCommandLine to a default. No lock taken since we're in the constructor so nothing can race.
-            ReparseCommandLine_NoLock();
+            // Set up _commandLineArgumentsForCommandLine to a default. No lock taken since we're in
+            // the constructor so nothing can race.
+            ReparseCommandLineIfChanged_NoLock(commandLine: "");
         }
 
-        public string CommandLine
+        /// <returns><see langword="true"/> if the command line was updated.</returns>
+        private bool ReparseCommandLineIfChanged_NoLock(string commandLine)
         {
-            get
-            {
-                return _commandLine;
-            }
+            var checksum = Checksum.Create(commandLine);
+            if (_commandLineChecksum == checksum)
+                return false;
 
-            set
-            {
-                if (value == null)
-                {
-                    throw new ArgumentNullException(nameof(value));
-                }
+            // Dispose the existing stored command-line and then persist the new one so we can
+            // recover it later.
 
-                lock (_gate)
-                {
-                    if (_commandLine == value)
-                    {
-                        return;
-                    }
+            _commandLineStorage.Dispose();
```

Comment:
On the first call from ctor, do we dispose the storage that we just allocated and allocate a new one? Would it make sense to keep `_commandLineStorage` uninitialized until this point?

Reply:
Also, creating a storage for an empty string seems unnecessary.

Reply:
So, i wanted to avoid that complexity :)   However, you @sharwell and @ryzngard all seem to want that, so i'll see if i can do it with low complexity :)

Reply:
I don't think we need to always avoid writing an empty string, but the initialization in the constructor could easily be removed. I'll let others speak for what they want though :)